Add new Package USER-RANN with pair style rann for using a neural network to compute energies and forces

kipbarrett opened the pull request:
**Summary**

Pair rann is a pair style which uses a feed forward neural network to compute atom energies and forces. We are requesting that it be included as a new USER package in LAMMPS. 


**Related Issue(s)**

N/A

**Author(s)**

Lead developer:             Christopher Barrett <barrett@me.msstate.edu
with contributions by      Doyl Dickel <doyl@me.msstate.edu>

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

Should be fully compatible.

**Implementation Notes**

1) Fingerprints and Activations are designed with a modular framework based on the way pairs, atom styles, fixes, computes, etc. are used in LAMMPS. This is so that more types of fingerprints and activations can easily be added later. 

2) We designed some fingerprints to interact with the SPIN package. The non-spin fingerprints and the rest of the code are fully functional without the SPIN package enabled. If spin fingerprints are used, the code will ensure that atom style spin is being used. If the SPIN package is not installed, the existing code already prevents atom style spin from being enabled.

3) We provided four fitted potentials in the potentials folder for examples, and detailed documentation.

**Post Submission Checklist**

<!--Please check the fields below as they are completed **after** the pull request has been submitted. Delete lines that don't apply-->

- [ ] The feature or features in this pull request is complete
- [ ] Licensing information is complete
- [ ] Corresponding author information is complete
- [ ] The source code follows the LAMMPS formatting guidelines
- [ ] Suitable new documentation files and/or updates to the existing docs are included
- [ ] The added/updated documentation is integrated and tested with the documentation build system
- [ ] The feature has been verified to work with the conventional build system
- [ ] The feature has been verified to work with the CMake based build system
- [ ] Suitable tests have been added to the unittest tree.
- [ ] A package specific README file has been included or updated
- [ ] One or more example input decks are included

**Further Information, Files, and Links**

<!--Put any additional information here, attach relevant text or image files, and URLs to external sites (e.g. DOIs or webpages)-->

akohlmey commented:
@kipbarrett thanks for your submission.
I have not looked at this too closely, but from what I saw, there will be some requests for moving files and some changes in the organization of the implementation coming. Some trivial, some not so trivial, and some where we need to have some internal discussions among the core LAMMPS developers first. Since we are currently busy with preparing the next patch release, this will have to wait a while before it we can do all of this. I will try to get a first round of (mostly cosmetic) suggestions out to you as quickly as possible, since it will probably take some time to get all of those sorted out.

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+}
+
+bool Fingerprint_bond::parse_values(char * constant, char * line1){
+	char **words=new char *[MAXLINE];
```
since we moved to C++11 as minimum C++ standard requirement, we are also switching away from some of the C-string string functions and use the C++ std::string class instead. Comparisons become more readable, search and substitution requires less code and functions that are not thread-safe like strtok() can be avoided. strtok() in particular is something that we want to get rid of and not add new uses to the code.
For parsing/processing text files there are utility functions in the utils:: namespace and particular there are the Tokenizer and ValueTokenizer classes that are already applied to a bunch of parsers for reading potential files.

kipbarrett replied:
I've removed all uses of strtok and used the Tokenizer instead. I did not use the text file reader because it has the line limit of 1024 hard coded. This conflicts with my potential file convention which has formatted matrices. The weight matrices can have more than 60 columns for alloys which with 16 digit precision will exceed that line length.

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+	else if (strcmp(constant,"m")==0){
+		m = strtol(words[0],NULL,10);
+	}
+	else pair->errorf("Undefined value for bond power");
```
This kind of error condition should either throw an exception or directly call `Error::one()` or `Error::all()`, you should be able to access those via `pair->lmp->error->`.

kipbarrett replied:
I have tried to make all the fingerprints and activations rely only on things defined by pair_rann. This is because I want to use the same code with minimal changes in my calibration routines outside of LAMMPS. For that reason, I put a wrapper around the error function, created a struct with the NeighborList information, etc. I've modified the error function now to pass the correct line and file to LAMMPS Error::one().

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+//Generate table of complex functions for quick reference during compute. Used by do3bodyfeatureset_singleneighborloop and do3bodyfeatureset_doubleneighborloop.
+void Fingerprint_bond::generate_exp_cut_table(){
+  int m,n;
+	double r1;
```
there is inconsistent indentation here. as a reminder, in LAMMPS indentations are 2 spaces per level, no TABs, and no whitespace at the end of a line.

akohlmey commented on src/fingerprint_bond.h:
```diff
@@ -0,0 +1,80 @@
+    int **coeffx;
+    int **coeffy;
+    int **coeffz;
+    int k;
```
please do not use class members with such trivial names and particularly not characters that are common to be used for loop indices (i,j,k,l,m,n)

kipbarrett replied:
Done.

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+	}
+}
+
+int Fingerprint_bond::factorial(int n) {
```
we have an optimized `factorial()` function in the MathSpecial namespace. Please use that one.

kipbarrett replied:
Done. I created a wrapper for it in pair_rann.

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+			kk=jj;
+			if (ktypes == nelements || ktypes == jtype){
+				count = startingneuron;
+				double dot = (y[jj][0]*y[kk][0]+y[jj][1]*y[kk][1]+y[jj][2]*y[kk][2]);
```
Please note that there are a whole set of functions for operations on 3-vectors in the MathExtra namespace.

kipbarrett replied:
This is the most computationally intensive part of the code. I think adding function calls for clarity is not worthwhile here.

akohlmey commented on src/pair_rann.h:
```diff
@@ -0,0 +1,171 @@
+#ifndef LMP_PAIR_RANN
+#define LMP_PAIR_RANN
+
+#define MAXLINE 1024
```
the define and the following include files should be moved from the header to the .cpp file. only the essential include statements like for pair.h should remain. Please see https://github.com/lammps/lammps/blob/master/doc/include-file-conventions.md for some explanations.

kipbarrett replied:
Done.

akohlmey commented on src/pair_rann.h:
```diff
@@ -0,0 +1,171 @@
+  void init_list(int , NeighList *);
+  void errorf(const char*);
+  int count_words(char *);
+  //black magic for modular fingerprints and activations
```
this should be done differently. please see the prototype code provided elsewhere.

kipbarrett replied:
I've made some changes to this which resolve the conflicts with the two build systems. I haven't done it exactly the way you suggested though. The reason why is that I would prefer that adding fingerprints/activations require as little modification of existing code as possibly. Reusing the LAMMPS style approach was a good way to do that. 

I agree that there is no reason to have LAMMPS build the style files on each compile. Instead, I suggest that developers can simply add new fingerprints/activations manually to the list of headers in rann_list_*. Changing these names to not use style_* fixes the build conflicts. Hopefully this approach is ok.

akohlmey commented on src/pair_rann.h:
```diff
@@ -0,0 +1,171 @@
+	bool **weightdefined;
+	bool **biasdefined;
+
+	struct Simulation{
```
this and the following struct do not need to be define here. a forward declaration should be sufficient. the full struct can be defined in the implementation.

kipbarrett replied:
I tried to get this to work without success. I was able to compile pair_rann.cpp with it defined as you suggest, but the fingerprint files would not compile since they also reference the structs. Maybe I'm not doing it correctly.

akohlmey commented on src/pair_rann.cpp:
```diff
@@ -0,0 +1,1247 @@
+}
+
+void PairRANN::errorf(const char * message){
+	this->error->all(FLERR,message);
```
please note that `Error::all()` may **only** be used for cases where **all** MPI ranks call it at the same time. Otherwise, `Error::one()` must be used or else LAMMPS will "hang".

kipbarrett replied:
I've changed all calls to Error::one().

akohlmey commented on src/pair_rann.cpp:
```diff
@@ -0,0 +1,1247 @@
+
+
+//treats # as starting a comment to be ignored.
+int PairRANN::count_words(char *line){
```
there is no need for a custom function here. there are several convenience functions for parsing and custom string processing in the utils namespace and you may also switch to use the Tokenizer/ValueTokenizer class for even more efficient and compact processing of parameter files.

kipbarrett replied:
I removed this function.

akohlmey commented on src/pair_rann.cpp:
```diff
@@ -0,0 +1,1247 @@
+//	firstneigh = listfull->firstneigh;
+//	jlist = firstneigh[i1];
+	int L = net1.layers-1;
+	double layer[net1.maxlayer];
```
these "fortran style" variable length array definitions on the stack are not acceptable and not compliant with C++ standards. They are accepted by GNU C++ compilers only in standard mode with accepts GNU extensions.
Since `jnum` can be large, they must not be allocated on the stack in the first place. Most Linux installations have limits on the stack anyway and thus would cause crashes. Please see the Memory class with its create(), grow(), and destroy() functions. You should also set things up so you avoid frequent re-allocations by storing a `jnummax` variable and then call "memory->grow()" only when jnum is larger than jnummax. there are plenty of example uses of this code construct elsewhere in LAMMPS.

kipbarrett replied:
I've changed all of these to use the LAMMPS memory class. Unfortunately this seems to slow computation by more than 10% compared to the non-compliant method.

akohlmey commented on src/fingerprint_bond.cpp:
```diff
@@ -0,0 +1,917 @@
+			}
+			if (atomtypes[1]!=atomtypes[2]){//Bb!=Bg
+				double Bg[mb];
+//				double dBgx;
```
please remove all commented out code. if there is debug code that should remain, please place it into
```
#if DEBUG_RANN
// debug code for XXX
[...]
#endif
```
and then place a `#define DEBUG_RANN 0` in a prominent position at the beginning of the file.

akohlmey reviewed (changes requested):
As promised, here is a first round of suggestions, mostly of the cosmetic kind, but also several related to programming style and code layout and proper integration into LAMMPS.

This should become a package USER-RANN and thus all files need to be
moved to src/USER-RANN, the package must be added to the conventional
make and the CMake based build system, the package must be documented
in the relevant places in the manual, the docs for pair style rann should
mention the need to install the package under Restrictions, and there
should be 1-2 small example inputs/logs in the examples/USER/rann folder.

There should be a new namespace RANN inside of LAMMPS_NS for all
functions and classes inside the USER-RANN package that are not the
pair style. Correspondingly the "fingerprint*.*" and "activation*.*"
files should be renamed to "rann_fingerprint*.*" and "rann_activation*.*,
respectively.

Please only use C-style stdio for informational output and only use
the predefined "screen" or "logfile" FILE pointers. Unless desired
otherwise, should should protect this from producing multiple output
when running in parallel by bracketing this with `if (comm->me == 0)`
For formatted output using sprintf() to a buffer should be avoided
and instead either fmt::format() or fmt::print() can be used.

Please note that for and if statements should be formatted as follows:
```
for (init; condition; increment) {
  ...
}

if (condition) {
  ...
}
```
i.e. with `) {` and not `){`. If the body is a single line,
the curly braces may be omitted (better not if there also is an else)
and if the line is very short (e.g. a break or continue statement,
or a single assignment) the body may be on the same line.

The use of this-> should be avoided where possible, and if this is
required for disambiguation, the variable or parameter names should
be renamed. You also should avoid having local variable names
"shadow" class member names.

The re-use of the implementation strategy of the factory pattern for
pair styles and so on is a very nice idea, but a) it would only make sense
to do it in this style if people would want to include/exclude those
style all the time and b) if there was the corresponding support in the
two build systems. As it is, this pattern collides with some of the
assumptions made in the build process, and thus we better find a way to
do the same thing without the preprocessor magic. At the link below you will find a minimal demo implementation that showcases how this could look like.

[factory-demo.tar.gz](https://github.com/lammps/lammps/files/5901108/factory-demo.tar.gz)

You will find a whole bunch of individual suggestions attached to individual
locations in the added source files.

kipbarrett commented:
Hi Axel,

Thanks for the quick response. I'll try to get these changes tended to over the next few days.

Best regards,

Christopher D Barrett
Assistant Professor
Department of Mechanical Engineering
Phone: 662-325-7325
Email: barrett@me.msstate.edu<mailto:barrett@me.msstate.edu>
Website: http://www.me.msstate.edu/people/christopher-barrett/<http://www.me.msstate.edu/people/christopher-barrett/>
Office: 213 Carpenter Hall
Mississippi State University
________________________________
From: Axel Kohlmeyer <notifications@github.com>
Sent: Sunday, January 31, 2021 5:45 PM
To: lammps/lammps <lammps@noreply.github.com>
Cc: Barrett, Kip <barrett@me.msstate.edu>; Mention <mention@noreply.github.com>
Subject: Re: [lammps/lammps] Pair rann (#2570)


@akohlmey requested changes on this pull request.

As promised, here is a first round of suggestions, mostly of the cosmetic kind, but also several related to programming style and code layout and proper integration into LAMMPS.

This should become a package USER-RANN and thus all files need to be
moved to src/USER-RANN, the package must be added to the conventional
make and the CMake based build system, the package must be documented
in the relevant places in the manual, the docs for pair style rann should
mention the need to install the package under Restrictions, and there
should be 1-2 small example inputs/logs in the examples/USER/rann folder.

There should be a new namespace RANN inside of LAMMPS_NS for all
functions and classes inside the USER-RANN package that are not the
pair style. Correspondingly the "fingerprint*." and "activation."
files should be renamed to "rann_fingerprint." and "rann_activation.*,
respectively.

Please only use C-style stdio for informational output and only use
the predefined "screen" or "logfile" FILE pointers. Unless desired
otherwise, should should protect this from producing multiple output
when running in parallel by bracketing this with if (comm->me == 0)
For formatted output using sprintf() to a buffer should be avoided
and instead either fmt::format() or fmt::print() can be used.

Please note that for and if statements should be formatted as follows:

for (init; condition; increment) {
  ...
}

if (condition) {
  ...
}


i.e. with ) { and not ){. If the body is a single line,
the curly braces may be omitted (better not if there also is an else)
and if the line is very short (e.g. a break or continue statement,
or a single assignment) the body may be on the same line.

The use of this-> should be avoided where possible, and if this is
required for disambiguation, the variable or parameter names should
be renamed. You also should avoid having local variable names
"shadow" class member names.

The re-use of the implementation strategy of the factory pattern for
pair styles and so on is a very nice idea, but a) it would only make sense
to do it in this style if people would want to include/exclude those
style all the time and b) if there was the corresponding support in the
two build systems. As it is, this pattern collides with some of the
assumptions made in the build process, and thus we better find a way to
do the same thing without the preprocessor magic. At the link below you will find a minimal demo implementation that showcases how this could look like.

factory-demo.tar.gz<https://secure-web.cisco.com/18qvd1yUH8IvWulQCK-K_R4URB0GDIBvrx5B5PX4Sqq7BOd0lmkK0IGlqmz8cL3lxNCg_MXQwJORfxFJYghcdsqDAHRj_QpvGBgkFm6VWAYLtkNY7alHZObCievxeIfAHd80n0obu44lLVLNGfGWzbnaSOiGqXMSNuc_fvvwUmKMUp6w1tAoJ5qU8lljuQcCopXt5zwLqpMA88ZubpzX5Q3_XEyfl1Kp-pgH9wK0sQg4bAugnd9ObZAPsTyevNK5zuVOMEQyckS0spI9NkHS29nSHZcChU3BlRgBiKgn7nwE-zvohu_T4HrNC9pWZtCtF/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Ffiles%2F5901108%2Ffactory-demo.tar.gz>

You will find a whole bunch of individual suggestions attached to individual
locations in the added source files.

________________________________

In src/activation.cpp<https://secure-web.cisco.com/1BJGGKJUgswezgNLPh7cCJpZv6je0LYPn1cQPsUcF1flrxsiY40QnVErnOUSfdEcY9IwvrUWKm4lyWmUXyFWP7wvYDWY3_CjOh40jRmgu04ZVkngfTv7kgWKNJSGqSk0YMnoJr1J4ZJFzHExA0YISuCB82fJrktSRpMJAwhUPiudnySXRNFyjxlLohGEFN6cnMov5TuRd6EpjgwIUI_dJHS2Qh5CV9hFrJD56hlu08EqpesrSjQTNGXyYTqr5PP3a_iXYCYpbIheR0nGY7RtLrEoboj9IZWKD-Locazw8vv1wFZ9ZZ3H_FcBQdKN66zXi/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567495263>:

> +#include "activation.h"
+
+
+using namespace LAMMPS_NS;
+
+Activation::Activation(PairRANN *pair) {
+       empty = true;
+       style = "empty";
+}
+
+Activation::~Activation(){
+
+}
+
+//default is linear activation (no change).
+double Activation::activation_function(double A){


to reduce overhead, this function should be define in the header file, so the compiler can inline it.

________________________________

In src/activation.h<https://secure-web.cisco.com/1_3MGIvf9b6g3XLao4ogrCJ-VjnEztRllprhAuyHbIVa8I3RtKv3aIognlsUm7JRUAtiK8qBsVE3O0rqhYXTaavjg02stGDsmLlFzcvXOLeOre2ODMKNEIBbTJqY3nz0-Q5Eo9a-x3NPl-kUNt_wWxmNd71FjIltG9-s-X6v08MwRDtF33AgOYswKX_QMtxM6v2zm3YPFFZYiJbJRyiHhlO5H7oQ28vdqbfS0piO9tnTyhMwGuz7vExca3GVbxr0qxVxhxRYzzTgGnzMit8XbIgVwRHsacMZaIkZQWYukEBNjJlmQEJRQAUGzA1FE01gN/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567495448>:

> +    ----------------------------------------------------------------------*/
+/*
+“The research described and the resulting data presented herein, unless
+otherwise noted, was funded under PE 0602784A, Project T53 "Military
+Engineering Applied Research", Task 002 under Contract No. W56HZV-17-C-0095,
+managed by the U.S. Army Combat Capabilities Development Command (CCDC) and
+the Engineer Research and Development Center (ERDC).  The work described in
+this document was conducted at CAVS, MSU.  Permission was granted by ERDC
+to publish this information. Any opinions, findings and conclusions or
+recommendations expressed in this material are those of the author(s) and
+do not necessarily reflect the views of the United States Army.”
+
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#ifndef ACTIVATION_H_


in LAMMPS the include guards follow the scheme LMP_<FILE>_<NAME>_H for a file called file_name.h
please adjust.

________________________________

In src/activation.h<https://secure-web.cisco.com/1o5jjXd_u0hN89Nm76_2GXg71BhBeSWTXC0LgsCGo01JTIaTZRgf5sRsFvcjoiWPly77i2gF_4OwTTUZJWHensXOI6uuBTn6_sUqchdscB46imjesJV_hQsIsolb7iD99mR0Z_tijENVslkgZOhE-ZGZBlH2VN8K6v5mxWgWchQ95JpRYXP1KmITATXPGj-nSQtRGrEXry1sJkC_lJhipWi0Kok10zaTpWU9Ahs9cw5uf6h5Cb9jG706aBIAoiOM4glDJYYweT56AKEuW8_SgQtFNySgiH5w9iH1rc7Itggr-00EB5kXZLjUhtY9UrLIS/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567495639>:

> +otherwise noted, was funded under PE 0602784A, Project T53 "Military
+Engineering Applied Research", Task 002 under Contract No. W56HZV-17-C-0095,
+managed by the U.S. Army Combat Capabilities Development Command (CCDC) and
+the Engineer Research and Development Center (ERDC).  The work described in
+this document was conducted at CAVS, MSU.  Permission was granted by ERDC
+to publish this information. Any opinions, findings and conclusions or
+recommendations expressed in this material are those of the author(s) and
+do not necessarily reflect the views of the United States Army.”
+
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#ifndef ACTIVATION_H_
+#define ACTIVATION_H_
+
+#include "pair_rann.h"


there is not need to include the header file here. A simple forward declaration of PairRANN will do.

________________________________

In src/activation.h<https://secure-web.cisco.com/1izbrE6wUZQrdlgCVS_Z_AKI8mlBEUlrFe72_MGjeBq4BbIJc44dxU-whaKCTi8dkkOTN20qffeFNAcYImSFFh2mgSeEdZa6-Iy4sHW-dNrCQv90EwnN3cJpo9eNcIK05F_Bn86mrMmUe_uJuZCHwtEVYc3hnmTmmDd8gMx1veMvEnLE90Nwh-OdV2dRMCQpVqTq9MVOIKWkpe2v4RTYUNr73C3Ub7uoj4omTwzPEphAsYU7D5d_s6yBGP2X7r4OORb9qN9yYPdG6uYjoGFNB5BWMAQmdMKc4dSZ4k1DCWAh9GiLNLfI5_2pZrCuGiao8/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567495682>:

> +this document was conducted at CAVS, MSU.  Permission was granted by ERDC
+to publish this information. Any opinions, findings and conclusions or
+recommendations expressed in this material are those of the author(s) and
+do not necessarily reflect the views of the United States Army.”
+
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#ifndef ACTIVATION_H_
+#define ACTIVATION_H_
+
+#include "pair_rann.h"
+
+namespace LAMMPS_NS {
+
+       class Activation {


indentation in LAMMPS is 2 spaces per level.

________________________________

In src/activation_linear.cpp<https://secure-web.cisco.com/1I1-uCajkcLJovI1_oYhOQ81sHnbD3GvxpL3RjxxtvhwSu8qls_1dwsNXto1KYrM4GCSZ4UbLDXODcVQWvonrZSwLcSJESI8AnD1y8lS0jpQiFPJc-vWBayPGKc7-7kJSIUt7l89WqXTjyWmw-5JKZIeG8qpqFdihI95n3ZSphw-LHwJNVVrJedSnPLmnULEMF7e-LQNanHmII2Q4jdT2qOBvQkOzLLV2Ygh5WK-WQ8fqQJlx72o600_YiHplWU7UH1q1jcQVRO8JY8AM1_HuE05c9fIwoY-YXPqv5L-vrvr856_obHxUZdfklEJkDG4v/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567497392>:

> +
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#include "activation_linear.h"
+
+
+
+using namespace LAMMPS_NS;
+
+Activation_linear::Activation_linear(PairRANN *pair) : Activation(pair){
+       empty = false;
+       style = "linear";
+}
+
+double Activation_linear::activation_function(double A)


same as for the Activation class, these should be in the header for efficiency. in fact, this could be a header-only class.

________________________________

In src/activation_sigI.cpp<https://secure-web.cisco.com/1UWb61J1Gt-e5NlP5Dq7hscsLuIEC6jjEqhgACwXw7W0m1VCbqk3mIh3j3eY0Lkz6NNu_YUolug9-xG6utQ6f72ka0lkmyBEUIv7nt28Sfy7ES0omgXiQwWvKK_DG7h8FcfH6EPjcQNtXS0qJ_OZO5e9uNuArCRKHat8zJICraZkcWho2zuXWkLif2AK9FAVGuzeiZBOsLth3xsULf58KXqyYVWA2PFe87ubCgEiLceGgbG6Cf5TOD35QoT9jWqCjAhTlu9Q4grQf7FV0BEXVnBFmg2QajWueYUZt-3l1cb4klSpE2LVTD4T85w6O6e9u/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567497514>:

> @@ -0,0 +1,52 @@
+/* -*- c++ -*- ----------------------------------------------------------


LAMMPS filename should have all lower case names. So please replace sigI with sig_i.

________________________________

In src/fingerprint.cpp<https://secure-web.cisco.com/1Qm884cS1_8mcPCw2RujjLzP3QswW1GNRZNAYgvba0v7w9qedwIXemRqngwY06TB0EfSEs9K37wNHSj14HKB2Ig5xg3KK2GAq_xirD7J0fmzfXSxOd99hPtbVqyPYPHZExxAUZ1Ye9q4IwchdLWu82qdAubPHLvjOoH8uB0Dfq2HjIJm0HkaKTxHBvwsOUXI4LPO319_tv5p3tVvp_P700BNpHmliw0DnaTxeB7ASo-sJ1KZTUAvlAVNZntbFYgFB28U00uVSOhBuPFI8tJt0uThqPqHNEEd8HyHa0c-QbXFTRhsfZZmB3iyVkLzQbd3o/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567497932>:

> +managed by the U.S. Army Combat Capabilities Development Command (CCDC) and
+the Engineer Research and Development Center (ERDC).  The work described in
+this document was conducted at CAVS, MSU.  Permission was granted by ERDC
+to publish this information. Any opinions, findings and conclusions or
+recommendations expressed in this material are those of the author(s) and
+do not necessarily reflect the views of the United States Army.”
+
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#include "fingerprint.h"
+
+
+using namespace LAMMPS_NS;
+
+Fingerprint::Fingerprint(PairRANN *pair)


better to use (PairRANN *_pair) as argument and then pair = _pair; below instead of forcing disambiguation through use of "this" below.

________________________________

In src/fingerprint.cpp<https://secure-web.cisco.com/1XJbASuF4R8KdyddG3iKQaVeh6wlhKLu6pXRqzbkmpRg71M8BbTh2oRu12YlzdDEvOhcL14c09r21linDALOesBrD7tDCE3wnS89uxFeZ27O0hi2n6J1R8pLIxfapUM1mebKreeKIYelxC89J85ITU6xtZILI8Bjw7BhW9MIBEDnqFkmdNWr6ncxFdbP0NpPvMBEiWWWYgicG7Ty6pbdo8-Dpf77C4qH0RSCRyggb4tEx3KqCgqNZADGCU9p2qK4MKYU-c-uDmKeEZ5O4AAEqxp4S-N7PSyuBu9THJ2buI2Udq3Hj8hdIeowB3x6RLyes/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567498322>:

> +
+void Fingerprint::write_values(FILE *fid){
+
+}
+
+int Fingerprint::get_length(){
+       return 0;
+}
+
+//Smooth cutoff, goes from 1 to zero over the interval rc-dr to rc. Same as MEAM uses. Used by generateradialtable and generatexpcuttable.
+double Fingerprint::cutofffunction(double r,double rc, double dr){
+       double out;
+       if (r < (rc -dr))out=1;
+       else if (r>rc)out=0;
+       else {
+               out = pow(1-pow(1-(rc-r)/dr,4.0),2.0);


please note that there is a much optimized (and more accurate) version of the power function for integer arguments, powint(), in the MathSpecial() namespace. for powers of 2, and 3 there are also square() and cube().

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1YODw7PcLTLd1RJEIN73OVLh8zJ7wgjL92zL0ukM6W9pWw4_n4ij-Uyq2vG-7-nHkhluJ956t250Zz07Mr5YkN3M4ZC_yHsqVcuCQeNCHvc7tRrBrOW6nCBT3lGw7zbSy3VS_UKNVgjHAFGpQoHWy0zhW8K1CrYsl18O0k4QycQoE2eSFJUcSN112EvJsP96YXs9g1J0aZBIWDvSkbvDCktsPRD2bAyOdM8lhFf4Jwm11c93mPC67iwm0E90fkgc6-mwSINmhfVmhcKwNM4h76kEeFHiOuPvojtvJOycSis3yexnZ3w4vIKWwt_Qojlzq/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567499002>:

> +             delete [] coeff[i];
+               delete [] coeffx[i];
+               delete [] coeffy[i];
+               delete [] coeffz[i];
+               delete [] Mf[i];
+       }
+       delete [] coeff;
+       delete [] coeffx;
+       delete [] coeffy;
+       delete [] coeffz;
+       delete [] Mf;
+       delete [] rinvsqrttable;
+}
+
+bool Fingerprint_bond::parse_values(char * constant, char * line1){
+       char **words=new char *[MAXLINE];


since we moved to C++11 as minimum C++ standard requirement, we are also switching away from some of the C-string string functions and use the C++ std::string class instead. Comparisons become more readable, search and substitution requires less code and functions that are not thread-safe like strtok() can be avoided. strtok() in particular is something that we want to get rid of and not add new uses to the code.
For parsing/processing text files there are utility functions in the utils:: namespace and particular there are the Tokenizer and ValueTokenizer classes that are already applied to a bunch of parsers for reading potential files.

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1frU9SPke9zhP_SuawFbm-mgPUO6AfJnlbN5rVmgtuYf8AkdGMUrXiXkjAR7WDQAvidlyV_cnopNZEmyOfbnVMdnb1eJJzbOJ562bJ2JoOD3FirdTt-aJsMgVkYGwoEgcWlvuSoEmT-Btmt_iCCZUAwBSe2FtjZ2m25bFkkhHEAAnzBmQGLOrqYpIeeW1b29PHB0xvfXSytXNUfSoAoLJHn1vILsWWSWqJMO02H2JnwwGxOcFVFFeBL-TlfWXeN0_x_B4IkiboaGvLDAYEk4m-Yc2UW4fTJmAjOHbHUvOuBmfYq9cYMn6LItkRm1jyW_w/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567499399>:

> +             delete [] alpha_k;
+               alpha_k = new double [nwords];
+               for (l=0;l<nwords;l++){
+                       alpha_k[l]=strtod(words[l],NULL);
+               }
+       }
+       else if (strcmp(constant,"dr")==0){
+               dr = strtod(words[0],NULL);
+       }
+       else if (strcmp(constant,"k")==0){
+               k = strtol(words[0],NULL,10);
+       }
+       else if (strcmp(constant,"m")==0){
+               m = strtol(words[0],NULL,10);
+       }
+       else pair->errorf("Undefined value for bond power");


This kind of error condition should either throw an exception or directly call Error::one() or Error::all(), you should be able to access those via pair->lmp->error->.

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1X5RnhPbQbajL9OjNXOJv3GacR0tsPopai3h4lQpBppqpIsGpiVox5iSCcvIMYcUcWd8aHwX3gzgfGgPpOYnmpCHXtxa1FMbH_5I1lN2RfQOF_XKQjhFgFv9AoB2XElwiKiuoZ8x7k2FD8BeNtzuchkHNWhR-YJrJuRmHl8apM8aETl5MPBFpmh9DkpxFDcRKzYmD9DEjnO_xRYT8tqSnQnK5jwwDoX3o9JAWxjJ9mcQFn6peBjYEQKcyf6HnB_pkKs_I96jX-lBQ4U2Y84S5oHj_NVlN0AuU8yGDUnyTWKHFJmRT5-S1FgiJWUegt5cq/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567499668>:

> +
+//number of neurons defined by this fingerprint
+int Fingerprint_bond::get_length(){
+       return m*k;
+}
+
+void Fingerprint_bond::allocate(){
+       generate_exp_cut_table();
+       generate_coefficients();
+       generate_rinvssqrttable();
+}
+
+//Generate table of complex functions for quick reference during compute. Used by do3bodyfeatureset_singleneighborloop and do3bodyfeatureset_doubleneighborloop.
+void Fingerprint_bond::generate_exp_cut_table(){
+  int m,n;
+       double r1;


there is inconsistent indentation here. as a reminder, in LAMMPS indentations are 2 spaces per level, no TABs, and no whitespace at the end of a line.

________________________________

In src/fingerprint_bond.h<https://secure-web.cisco.com/1t3Zp-oaA9JAFJAUY0urvEXxZ8ouDLQJMuKKHGR5TbUHTHRjIYZwB_KuDdV8ZqwtxlL6RW_4MstVT75bHCoDNTYgedQSJAdSNQN3SXqEBL8Dr-_7Cj0xfN7cCZHlRx8UPgyA-gPtsLupRm3ZpG968SYJfzTd5T0u4jTah9Si9hJWysfQmG50S_d9XxXmOFq9gtYvVMsib42WlM0S_qlVyigWIqWtTBAfOuu8i18gKOUeGHOWO4xOZVD1-isLtpCkT3fDik20khaPta_G3gSOc9ehv8LaIApt3Eh054_xaHrdhIylBR6bS9rzouoyovCOW/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567499855>:

> +     void do3bodyfeatureset_singleneighborloop(double *,double *,double *,double *,int,int,double *,double *,double *,int *,int,int *);
+       int factorial(int);
+       void generate_exp_cut_table();
+       void generate_coefficients();
+       int get_length();
+
+       double *expcuttable;
+    double *dfctable;
+    double dr;
+    double *alpha_k;
+    double re;
+    int **coeff;
+    int **coeffx;
+    int **coeffy;
+    int **coeffz;
+    int k;


please do not use class members with such trivial names and particularly not characters that are common to be used for loop indices (i,j,k,l,m,n)

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1ZsbqgY3AFIgLKhshZFnQOt2rbH8Hdl6_l1qU3ZnA2QF0DlncOOblA7XLPeNMCkIyHaVNk0qfO4KzN9g1r9v1RolF381MHBqGJ1CtlJ_Q9K4WdCh_3Lz-biAwtITy8zAzk8K0hPccoSPI-6kfJuW8GAp1ZgG874a9RyTZBYLIgMDA4PY4KQWGt0z3n9gj_v8cekP6XzAQ--A6guCBetaZs7d44f-dsPQkfjh98sX8dHzSmCPlTUUIayKD4VVMZTQYEhcV5MGF6ik9r85Dv3RPp0M-3PLDAwRkqhCI9NNf0dg-nknmNGDOE8o2iDlNrnjh/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567499985>:

> +     }
+       for (jj=0;jj<jnum;jj++){
+               if (expr[jj][0]==0){continue;}
+               count = startingneuron;
+               for (n=0;n<kb;n++){
+                       for (m=0;m<mb;m++){
+                               dfeaturesx[jnum*f+count]-=dfeaturesx[jj*f+count];
+                               dfeaturesy[jnum*f+count]-=dfeaturesy[jj*f+count];
+                               dfeaturesz[jnum*f+count]-=dfeaturesz[jj*f+count];
+                               count++;
+                       }
+               }
+       }
+}
+
+int Fingerprint_bond::factorial(int n) {


we have an optimized factorial() function in the MathSpecial namespace. Please use that one.

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1C6AaewUshVekZB2EW-T-6W7vkXS8V_0qHt2qNy2XW-VfUL-pRe1_z7Y8Eyes4uFwsGcP2i8v7WWzIwdNm0vWboL6pxptrMpFV6IcEIpnKfgdNRloTALtLPy3CdvIHG8sPnLGY8kzdDRF8YseY_D-P1GBKUFZRlZqQlVjlk83lg3Ja9suhh9JD9iAcXXVO2AJ_1WZkHxZb173Y4OpW62KfzjO_fS_V08kGrl2bInILBkcpLmGnwkPsS7bMjwUA8okaky9_StBDfWrG1baA16_dF10Q-XJ0Y3JnfXO5_PG8Wa6opMBTiNkrs62rnlFNg76/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567500256>:

> +                                             double c9 = dot1*(m*c3+c6);
+                                               dfeaturesx[jj*f+count]+=c7;
+                                               dfeaturesy[jj*f+count]+=c8;
+                                               dfeaturesz[jj*f+count]+=c9;
+                                               dfeaturesx[kk*f+count]+=dot1*(m*c10+c42);
+                                               dfeaturesy[kk*f+count]+=dot1*(m*c11+c52);
+                                               dfeaturesz[kk*f+count]+=dot1*(m*c12+c62);
+                                               dot1*=dot;
+                                               features[count++]+=2*dot1;
+                                       }
+                               }
+                       }
+                       kk=jj;
+                       if (ktypes == nelements || ktypes == jtype){
+                               count = startingneuron;
+                               double dot = (y[jj][0]*y[kk][0]+y[jj][1]*y[kk][1]+y[jj][2]*y[kk][2]);


Please note that there are a whole set of functions for operations on 3-vectors in the MathExtra namespace.

________________________________

In src/pair_rann.h<https://secure-web.cisco.com/1Cx6uIIxCaZy4y9_60rkFi9UIX8yvL2GO2j6TZonoOY4TxnogQTjerSnLvvH3Knchh_M8DwcRvN7D3Do6AjcfZRlgx_BxGCkMcr2LOTGWBTsl8-YjEGHkOJUGnSGYJGOgY5Et-F0-Y3_uAwKuPVMT376xrXTBgM-1Lspvx5lRTvuJg3voz5A1q6oSjHZYU0NiXjUZrLIGSgIpdoDp6Fvu0f3sKNROD7VCdHHQf_OFkBot92cioK3_tSeippahyQZZEdGUVgodoi7RPFZJsz5LKaXQIPSwU7YoPKmlY5lYA5WazO5zx9BARrW82LDkizIg/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567500645>:

> +recommendations expressed in this material are those of the author(s) and
+do not necessarily reflect the views of the United States Army.”
+
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#ifdef PAIR_CLASS
+
+PairStyle(rann,PairRANN)
+
+#else
+
+#ifndef LMP_PAIR_RANN
+#define LMP_PAIR_RANN
+
+#define MAXLINE 1024


the define and the following include files should be moved from the header to the .cpp file. only the essential include statements like for pair.h should remain. Please see https://github.com/lammps/lammps/blob/master/doc/include-file-conventions.md<https://secure-web.cisco.com/1QHFLCBWE-3t4EehDQQ_nicWxXhzBstsbxgCRr_1Dnqoqw_VtF4hKhKH99MVDG9WhFcvpQsIMljy_HJBJ87v9TL2t_T2TyPAy0DLMl0ino1q7DEwPUoWXGW2-hDyT9RHXpsovLtAq-ZS0F-psEJza6wVfsu9q_3xM2MNPLCGlHBuzGfMix4-rMEEnKUd1rGFhTBBnNYAP9_17HjzgqF1xKifK_ZYKrKvWhcIeUlXe4RtaBfIXvBlVXyr5PNWWeWdDrilzEXXQpTy7i0_vNfjL-D_1OoEVS_WfjQ23QqFlMlZK3IuqrMhSJHnv-QdQ_phs/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fblob%2Fmaster%2Fdoc%2Finclude-file-conventions.md> for some explanations.

________________________________

In src/pair_rann.h<https://secure-web.cisco.com/1ZtUzqAGdvccK3PkY6rAB54ABpoiLxGU6TNXolAyjse6m82ORvDTLa_9JQBzH8AIyhSbTyIPo-n1MwaGmNWnUerX930b7Ofs0reH_sPVFypfiQs-ViU_qOu_LOR71x5DdNmzQB0r94b_SZjgAutEVIxhWOJ3fmRzEkHo32SlGP_29UOLlvXLfovOEs_mGa8hJjbaFP7W-_orj9mTloK01hRRrk4NYSsAAWCWnL6pd0_n-Fz5ugUNCaQK6xPDmq72XLceOCA5zpDbl6YFoV41quQwqq5gCunJHyEYSR_-WvhUyMAOKMjt_oZlxTcdB9qBr/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567500778>:

> +
+class PairRANN : public Pair {
+ public:
+
+  //inherited functions
+  PairRANN(class LAMMPS *);
+  ~PairRANN();
+  void compute(int, int);
+  void settings(int, char **);
+  void coeff(int, char **);
+  void init_style();
+  double init_one(int, int);
+  void init_list(int , NeighList *);
+  void errorf(const char*);
+  int count_words(char *);
+  //black magic for modular fingerprints and activations


this should be done differently. please see the prototype code provided elsewhere.

________________________________

In src/pair_rann.h<https://secure-web.cisco.com/1W6SbmFolhd5A0waYuu5I14YaXihWQsM519OZhEklGKSdWBnUKQ1RMSI5elyHtg3DC1BlDxThTDcQrW_ggCMaTI3gnG8k899VeH0oEH26A4RrVF8W9y7DcNlVQvVErYOLrdGsKEyTW5dWNX7GxTloCiyVxkOFVdagJJi8ahWoltEcFb30jnG5dhKR3LqAajRMniSJ5ek1bv9q4vtCnA8Vh_FUeRl9oafksybBCsOBgWT2PyMrScHqIHXchDhYZNlDEwpi8yhwJLNWXTG7zkqQXUAS9vgfDhLYiMp7inVWxQCGy4q_m_fbZO4A6qYsr_uq/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567500923>:

> +  double cutmax;                             // max radial distance for neighbor lists
+  int *map;                     // mapping from atom types to elements
+  int *fingerprintcount;               // static variable used in initialization
+  int *fingerprintlength;       // # of input neurons defined by fingerprints of each element.
+  int *fingerprintperelement;   // # of fingerprints for each element
+  bool doscreen;//screening is calculated if any defined fingerprint uses it
+  bool allscreen;//all fingerprints use screening so screened neighbors can be completely ignored
+  bool dospin;
+  int res;//Resolution of function tables for cubic interpolation.
+  int memguess;
+       double *screening_min;
+       double *screening_max;
+       bool **weightdefined;
+       bool **biasdefined;
+
+       struct Simulation{


this and the following struct do not need to be define here. a forward declaration should be sufficient. the full struct can be defined in the implementation.

________________________________

In src/pair_rann.cpp<https://secure-web.cisco.com/1ZeRfKbX8ZNDzLpRHh3xo9K_UaJ7uJcSvL90dm1CCnjQYBr25wWbXU4FmZfGBnDmPsTlhXm2oedlJG1wnM3-sFwAMdT-gJOLUoGsZOtboFX0RdkK7BLgAW6AZQgnZdMGaDLEoByDcgCrIVTL1X2tLvLYFktCvfk_yyrm2qf5nKbjjBlyS15g7Ek3IKrnETiwau8jZCj3X0KWKIadqVwVkNqG7L28EZlGYQg9qH_WwNT3u6KMcNtKtrFm2WAzay8SOBvuaPRvEfO5tRX2Z473N8h-RJTkIA2wiE6tuC1YGXOCUIjiiV31B7rKVr78xZufA/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567501193>:

> +       neighbor->requests[irequest_full]->half = 0;
+         neighbor->requests[irequest_full]->full = 1;
+}
+
+
+/* ----------------------------------------------------------------------
+   init for one type pair i,j and corresponding j,i
+------------------------------------------------------------------------- */
+
+double PairRANN::init_one(int i, int j)
+{
+  return cutmax;
+}
+
+void PairRANN::errorf(const char * message){
+       this->error->all(FLERR,message);


please note that Error::all() may only be used for cases where all MPI ranks call it at the same time. Otherwise, Error::one() must be used or else LAMMPS will "hang".

________________________________

In src/pair_rann.cpp<https://secure-web.cisco.com/1QftXLWCKWjK6bVQ05fLMp5_74url8J-Q4l7_9X7Xadl-2dotOZophe8Up9AECqfYDWdpHgI5yTCpt1cBhwyuxHcUViaer0wDt1CDQL_dDU9n1t-t41v3rlKeMMiW5eGtM_Q3g-RMF6lK-Ye3aqGGD39bIjkZguZvy-agbwhLo5BO1BtncKCVzb2HbNslL6jUx_bwHrh-4rwVUHNOqZSJKfP8LBkdfS0d_emabr3Ivdo7WTS3SA0SDcRJtdfHnluZZtnlHE1i3TpV-rm4wAASY52mgPT0o6rtzEb6-KMKjrhkEu2CYVUx72UcGqycXi7c/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567501481>:

> +//                           if (Bij[jj]==false)continue;
+                               dlayerx[jj][j] = dlayersumx[jj][j];
+                               dlayery[jj][j] = dlayersumy[jj][j];
+                               dlayerz[jj][j] = dlayersumz[jj][j];
+                               dsx[jj][j] = dssumx[jj][j];
+                               dsy[jj][j] = dssumy[jj][j];
+                               dsz[jj][j] = dssumz[jj][j];
+                       }
+               }
+       }
+}
+
+
+
+//treats # as starting a comment to be ignored.
+int PairRANN::count_words(char *line){


there is no need for a custom function here. there are several convenience functions for parsing and custom string processing in the utils namespace and you may also switch to use the Tokenizer/ValueTokenizer class for even more efficient and compact processing of parameter files.

________________________________

In src/pair_rann.cpp<https://secure-web.cisco.com/1f1WCximBz39aHwMDKCECFaRDWKD58cb7B1bEVlNRr3_mDYQOTBFz4nuBYkEjizleLp401af7P1zfaRChd5FpEjIfw2J5nUOv4twwt76-67hipJUZKuTB7fLqnsGT-OdXXgLyXimqmhHfCK7VSwRrXdOY2L4OS1PJibtCz1gLCIcSMCna50z1TKsivAiN5LYB8h_DhiPqBdq4FgKX1UMm7QyIaquEUD4bN0hd5qPjh2Wqg2DYYxQoCsKqkkzvm_FE2lYTNLQAhN8YjE5TVjnY6erh6PmqZe-yr-zenIe0gM7l9K5kzrLceqV7XjcvknOw/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567502221>:

> +//Called by getproperties. Propagate features and dfeatures through network. Updates force and energy
+void PairRANN::propagateforwardspin(double *features,double *dfeaturesx,double *dfeaturesy,double *dfeaturesz, double *sx, double *sy, double *sz, double * energy,double **force,double **fm,double **virial, int ii,int jnum,int *jl){
+       int i,j,k,jj,j1,itype,i1;
+       int *ilist,*numneigh;
+       ilist = listfull->ilist;
+       int inum = listfull->inum;
+       int *type = atom->type;
+       i1=ilist[ii];
+       itype = map[type[i1]];
+       NNarchitecture net1 = net[itype];
+//     numneigh = listfull->numneigh;
+//     jnum = numneigh[ilist[ii]]+1;//extra value on the end of the array is the self term.
+//     firstneigh = listfull->firstneigh;
+//     jlist = firstneigh[i1];
+       int L = net1.layers-1;
+       double layer[net1.maxlayer];


these "fortran style" variable length array definitions on the stack are not acceptable and not compliant with C++ standards. They are accepted by GNU C++ compilers only in standard mode with accepts GNU extensions.
Since jnum can be large, they must not be allocated on the stack in the first place. Most Linux installations have limits on the stack anyway and thus would cause crashes. Please see the Memory class with its create(), grow(), and destroy() functions. You should also set things up so you avoid frequent re-allocations by storing a jnummax variable and then call "memory->grow()" only when jnum is larger than jnummax. there are plenty of example uses of this code construct elsewhere in LAMMPS.

________________________________

In src/fingerprint_bond.cpp<https://secure-web.cisco.com/1FJqsfr8oSXcmZshN2buxdIDOFUzifJNrr5pTq0GEAbj22lXL8ARakG5CwE7_O-EEm7N2MSk1kCVjb09PZ9a6otGQDHxivfeoWOHeXIGoLE3or9zfvEonHElHz1SkAk-nEr7TPsDUPMrXaXLeG6NYp6tSjb8PCviu1SBD2gOGFzlDhFTPnXGpA4kmZodfhtbQLlw3Ic1fZ7zYFe44gNfZ-krhNaSrNsrTsm9JCV5QtteqY6zXr8_9mnR2eLZsEZGgIVWag859_num75_Q6bkXnIpAHsGMKSrBN1BOtjAli6v5zaBAwsLewbMNG32AwdE4/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23discussion_r567502693>:

> +                             for (a2=0;a2<a;a2++){
+                                       yprod *= y4[M[a2+1]];
+                               }
+                               for (a2=a;a2<mb;a2++){
+                                       Bb[a2]=Bb[a2]+yprod;
+//                                     if (i==5 && a2 == 1 && n==0){
+//                                             char str[128];
+//                                             sprintf(str,"%f %f %f\n",Bb[a2],yprod,y4[M[a2]]);
+//                                             std::cout<<str;
+//                                     }
+                                       yprod *= y4[M[a2+1]];
+                               }
+                       }
+                       if (atomtypes[1]!=atomtypes[2]){//Bb!=Bg
+                               double Bg[mb];
+//                             double dBgx;


please remove all commented out code. if there is debug code that should remain, please place it into

#if DEBUG_RANN
// debug code for XXX
[...]
#endif


and then place a #define DEBUG_RANN 0 in a prominent position at the beginning of the file.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://secure-web.cisco.com/1fBpMJm9xaR8HdC-LK_O0DRjeYYQmYobI4TiFLDicOIDHliLoi-VhtmTup3VfA8Sn9zQbNqUXs7wGpc5fWrCimJRsv0CuzKb2Kqb1G-G409zyrMbPk9L78wiPo1A9pYzAVjlaqqm6eq-Z33jD5mhpEFrIqNhjwchMQtcrGijhNJj37TOEKGocMBScikF5L2tChDrC70fqxNep_FPjyjpOPafr_AC3Wn-QUToitQSl9kbq356yMnUUqsQ69b3-8S-mpMdIyxjCNzox-zRnOHu9NDU4T0J3HFC4q5sNvCaSE71WEWXO2eRbNnUp5ct2qGxT/https%3A%2F%2Fgithub.com%2Flammps%2Flammps%2Fpull%2F2570%23pullrequestreview-579942750>, or unsubscribe<https://secure-web.cisco.com/1rFFnILO1dTo8SYC2d_oLBOTHujbrLMd2VBZTTEnSGT5KPYwFqb-1dnZDGytNi2zrwSlXxzog27FqXh3kRHeqUyoNFo96779pDRfdLjQz0cw4J0KUtuTgTZ4MdL5ZtxIh96Ac5Gpa4MD2KxOh9bVpP4f6rRz3Pi86ujN_siz75PDTUGxXrEl8_G3L3xdlyizehowUMLSniDy7wbBtDU99Olc1tlyJHNhbN5pugKQngkCB4Emb8LIkSa-Sf9d-6Szt71rnBRNFezhIRKxev-smxMASSMam-dkgHyuBK-KdoWG3S26Bv-6U-zZO81936Ef6/https%3A%2F%2Fgithub.com%2Fnotifications%2Funsubscribe-auth%2FASTCKX2KRX44U7ITEEFHUE3S4XTSVANCNFSM4WZWR7WA>.

kipbarrett commented on src/activation.h:
```diff
@@ -0,0 +1,52 @@
+DISTRIBUTION A. Approved for public release; distribution unlimited. OPSEC#4918
+ */
+
+#ifndef ACTIVATION_H_
```
Done

kipbarrett commented on src/activation.h:
```diff
@@ -0,0 +1,52 @@
+#ifndef ACTIVATION_H_
+#define ACTIVATION_H_
+
+#include "pair_rann.h"
```
Done

kipbarrett commented on src/activation.h:
```diff
@@ -0,0 +1,52 @@
+
+namespace LAMMPS_NS {
+
+	class Activation {
```
I believe I have this resolved everywhere now.

kipbarrett commented on src/activation_linear.cpp:
```diff
@@ -0,0 +1,54 @@
+	style = "linear";
+}
+
+double Activation_linear::activation_function(double A)
```
I put all the activations as headers only. This should be generally appropriate for activations.

kipbarrett commented on src/activation_sigI.cpp:
```diff
@@ -0,0 +1,52 @@
+/* -*- c++ -*- ----------------------------------------------------------
```
Done

kipbarrett commented on src/fingerprint.cpp:
```diff
@@ -0,0 +1,113 @@
+
+using namespace LAMMPS_NS;
+
+Fingerprint::Fingerprint(PairRANN *pair)
```
Done.

akohlmey commented:
Thanks for your changes. Unfortunately, they address some problems, but also contain new problems. I am tied up with teaching for the next couple of weeks, so it will likely take a while before I can look at this again properly and provide feedback and more suggestions.

akohlmey commented:
@kipbarrett after some thinking we will insist on removing that convoluted factory code with complex pre-processing and a function pointer map and replace it with something simpler. We have just done the same thing for the KIM package and in the process figured out a cleaner way to do this, and I am actively thinking about ways to do a similar thing to the code in the core of LAMMPS. It is in its current form mostly as an intermediate step for a better solution from a previous implementation that would break with some compilers.  Please find attached a revised demonstration of how this could be done instead. Please note that I removed the "empty" boolean as it is redundant. Its function could be replaced by checking whether the pointer to the instance is NULL or not. Please also note that the activation function should be a pure function in the base class as the base class should not be used.

[factory-demo.tar.gz](https://github.com/lammps/lammps/files/6037247/factory-demo.tar.gz)

akohlmey commented on src/USER-RANN/pair_rann.h:
```diff
@@ -0,0 +1,174 @@
+    void errorf(const char*,int,const char*);
+    int factorial(int);
+    //black magic for modular fingerprints and activations
+    class RANN::Activation ***activation;
```
no "class" needed here since there is already a forward declaration above.

akohlmey commented on src/USER-RANN/pair_rann.h:
```diff
@@ -0,0 +1,174 @@
+    NNarchitecture *net;//array of networks, 1 for each element.
+
+   private:
+    template <typename T> static RANN::Fingerprint *fingerprint_creator(PairRANN *);
```
these two lines should also be removed

akohlmey commented on src/USER-RANN/pair_rann.h:
```diff
@@ -0,0 +1,174 @@
+    template <typename T> static RANN::Fingerprint *fingerprint_creator(PairRANN *);
+    template <typename T> static RANN::Activation *activation_creator(PairRANN *);
+    //new functions
+    void allocate(std::vector<std::string>);//called after reading element list, but before reading the rest of the potential
```
passing containers by values are a *very* bad idea. This forces needless copies. Those should be const references.

akohlmey commented on src/USER-RANN/pair_rann.h:
```diff
@@ -0,0 +1,174 @@
+    void allocate(std::vector<std::string>);//called after reading element list, but before reading the rest of the potential
+    void read_file(char *);//read potential file
+    void strip_comments(char *,FILE *fp);
+    void read_atom_types(std::vector<std::string>,std::vector<std::string>);
```
are all of these functions really needed as class members? couldn't those just be static functions in the class implementation that take a pointer to the data they should update?

kipbarrett replied:
Technically, no, they're not necessary as class members. However, static functions would require a lot of rewriting and I would argue it makes the code more convoluted and harder to read, especially since many of them call each other. I can't see that there is a strong reason to change this.

akohlmey commented on src/USER-RANN/pair_rann.cpp:
```diff
@@ -0,0 +1,1261 @@
+    Tokenizer values1 = Tokenizer(line1,": ,\t_\n");
+    linev = values.as_vector();
+    line1v = values1.as_vector();
+    if (linev[0].compare("atomtypes")==0)read_atom_types(linev,line1v);
```
why use std::string::compare()? it would be much more readable to use `if (linev[0] == "atomtypes") { `
also the lack of a blank between the conditional and the function calls makes this part of the file difficult to read.

akohlmey commented on src/USER-RANN/pair_rann.cpp:
```diff
@@ -0,0 +1,1261 @@
+  }
+}
+
+void PairRANN::read_atom_types(std::vector<std::string> line,std::vector<std::string> line1) {
```
why make this a function? this would be simpler (and better to follow) to just do in place. BTW: appending an item to a vector is simply done with `vector::push_back()`.

kipbarrett replied:
It is easier to read the potential file parser if all the pieces of the file are read in a consistent way. Thus, each section of the file has its own function and the read_file function can have a nice compact block of conditionals to check where to direct each pair of lines for parsing. This one is pretty short, but it is better to leave it alone. vector::push_back() has now been implemented.

akohlmey commented on src/USER-RANN/pair_rann.cpp:
```diff
@@ -0,0 +1,1261 @@
+  int nwords = 0,i;
+  for (i=0;i<nelements;i++) {
+    if (line[1].compare(elements[i])==0) {
+      mass[i]=strtod(line1[0].c_str(),NULL);
```
please use `utils::numeric()` for conversions of strings to numbers.

kipbarrett commented:
The factory code has now been replaced with the suggested hardcoded form. I've also dealt with most of the other comments.

kipbarrett commented:
@akohlmey and @sjplimp,

Thanks for looking over this and the pertinent suggestions. Hopefully this version is closer to being ready add to LAMMPS. @akohlmey, I'm not sure what you mean by a "Fortran feel". The code is written primarily to be as fast and flexible as possible for a neural network. I'm hopeful we can have it included soon since the new potentials we've developed for this are proving to work really well and we'd like to be able to share them more easily.

akohlmey commented:
@kipbarrett please let us know of your plans for this pull request. It currently fails to build for some configurations and has a merge conflict with cmake. If you need help or explanations or how to find the details of the build failures, please let us know.

Please note that we currently have multiple pull requests with neural network based pair styles pending and ideally we want to do the final review and processing of all of them for the same patch release so responses may be slow at times while we focus on ongoing refactoring work in the code base.

kipbarrett commented:
Hi Axel,

Some guidance on where the build failures occur would be helpful. I haven't found detailed logs for where it is failing on github, and on my machine it builds fine with both make and cmake. Also, I don't understand the merge conflict with cmake. The only thing I changed in cmake is adding the package name in the appropriate file.

akohlmey commented:
> Hi @akohlmey ,
> 
> Some guidance on where the build failures occur would be helpful. I haven't found detailed logs for where it is failing on github, and on my machine it builds fine with both make and cmake. 

the build failures are consistently with cmake only. most likely there is are/files that you have removed locally but they are not removed in the pull request branch that cause the conflict. the message is more-or-less the same. please check out the console log for one run here:
https://ci.lammps.org/blue/organizations/jenkins/dev%2Fpull_requests%2Fcentos7%2Fcmake_mpi_openmp_bigbig_static/detail/cmake_mpi_openmp_bigbig_static/2257/pipeline

> Also, I don't understand the merge conflict with cmake. The only thing I changed in cmake is adding the package name in the appropriate file.

There likely have been changes to upstream that produce the conflict. You could try doing a `git pull git://github.com/lammps/lammps.git master` and then resolve the resulting merge conflict and push that updated version back to github.

kipbarrett commented:
Thanks @akohlmey, 

That resolved all the build failures.

kipbarrett commented:
Hi @akohlmey ,

Could we get a status update?

akohlmey commented:
> Hi @akohlmey ,
> 
> Could we get a status update?

The pull request is still pending and I have not had the time to look into it. this is mostly because it will take a significant amount of time to get things properly checked and to look into better alternatives to elements that I believe can be improved.
Unfortunately, there is currently nobody else that has the time necessary to look into pull requests at that level and also we have a bunch of rather urgent projects that need to be completed before the stable release, the publication of a new LAMMPS paper, and the LAMMPS workshop this summer, so those are taking precedence.

Your pull request is not alone in that state. There are a few more on the pending list that have been sitting there even longer.

I understand that it is frustrating to see other pull requests going in while you are waiting on yours. Yet, there is little that can currently be done about it unless we have more people joining the core development team. In fact, if it had not been for the pandemic, we would be behind much more.

akohlmey commented:
@kipbarrett i've made some formal changes to the pull request to update it to the latest upstream version, resolve pull request, have it pass through integration testing and fix up some format and style issues. I will be looking into some code refactoring next, but I wanted to touch base with you before doing that.

There are two things that I would like your feedback on:

1) We have now an agreed upon "clang-format" configuration for source code in the "src" folder. this is not mandatory, but it helps keeping things consistent (even though not everything remains readable, so different parts need to be protected with '// clang-format off/on' comments. The default setting is to enable clang-format processing for .h files and only selectively for .cpp files. If you have clang-format (version 8 or later) installed on your machine, then please update your local branch of the pull request and try running 'clang-format -i <some .cpp file>' (after removing the '// clang-format off' comment at the top and see if you are comfortable with those changes or would prefer to keep the files as is. I have processed one pair for files for demonstration (`rann_fingerprint.cpp/.h`).

2) we have the convention of labeling potential files with useful information (units, date, authors, citation) as a comment in the first line and also check the UNITS: tag against the units setting in LAMMPS. I thing that would be a good thing for rann pair style potentials to follow. Related to that, I would also prefer to rename the potential to end in .rann instead or .nn to stick with the convention for most potential files.

Please let me know what your take on these issues are.

kipbarrett commented:
Hi @akohlmey 

I'm fine with both of these suggestions, as long as the clang formatting doesn't break anything. I'll update the potential files as soon as I can. Do you want me to do the clang formatting, or just do it yourself since you're already working in the code?

akohlmey commented:
> Hi @akohlmey
> 
> I'm fine with both of these suggestions, as long as the clang formatting doesn't break anything. I'll update the potential files as soon as I can. Do you want me to do the clang formatting, or just do it yourself since you're already working in the code?

It is trivial for me to do the conversion while editing files, since I have my editor configured to run it on a keyboard shortcut.
That also allows to quickly undo the change, if the result is not desirable.

I will generate reference output for the provided example and confirm that any changes (be it from clang-format, be it from my refactoring attempts) will not change the results. I will also add a tag to the current state to allow to roll back any changes to this state, since the current state fulfills the minimum formal requirements to be merged. I still need to review the code for functional issues, but that will be effectively included in the planned refactoring to take better advantage of the convenience functions and classes included in LAMMPS and avoid function calls and code constructs we want to get rid of where possible.

akohlmey commented:
@kipbarrett change of plans. After a discussion with @sjplimp and considering some "logistical issues" related to planned changes, we will be merging this PR as is today and then do a patch release later today. The refactoring, reformatting and other changes will then be proposed in a later pull request, *after* the reorganization of the LAMMPS source tree.

sjplimp commented:
@kipbarrett Thanks for contributing this to LAMMPS and working with us so it could be refactored to be more LAMMPS compatible.
